src/backend/allocator/dmabuf.rs

```diff
@@ -0,0 +1,177 @@
+    ) -> Option<Dmabuf> {
+        if offsets.len() < planes
+            || strides.len() < planes
+            || fds.len() < planes
+            || planes == 0
+            || planes > MAX_PLANES
+        {
+            return None;
+        }
+
+        let end = [0u32, 0, 0];
+        let end_fds = [0i32, 0, 0];
+        let mut offsets = offsets.iter().take(planes).chain(end.iter());
+        let mut strides = strides.iter().take(planes).chain(end.iter());
+        let mut fds = fds.iter().take(planes).chain(end_fds.iter());
+
+        Some(Dmabuf(Arc::new(DmabufInternal {
+            num_planes: planes,
+            offsets: [
+                *offsets.next().unwrap(),
+                *offsets.next().unwrap(),
+                *offsets.next().unwrap(),
+                *offsets.next().unwrap(),
+            ],
+            strides: [
+                *strides.next().unwrap(),
+                *strides.next().unwrap(),
+                *strides.next().unwrap(),
+                *strides.next().unwrap(),
+            ],
+            fds: [
+                *fds.next().unwrap(),
+                *fds.next().unwrap(),
+                *fds.next().unwrap(),
+                *fds.next().unwrap(),
+            ],
+
+            width: src.width(),
+            height: src.height(),
+            format: src.format(),
+        })))
+    }
+
+    /// Return raw handles of the planes of this buffer
+    pub fn handles(&self) -> &[RawFd] {
+        self.0.fds.split_at(self.0.num_planes).0
+    }
+
+    /// Return offsets for the planes of this buffer
+    pub fn offsets(&self) -> &[u32] {
+        self.0.offsets.split_at(self.0.num_planes).0
+    }
+
+    /// Return strides for the planes of this buffer
+    pub fn strides(&self) -> &[u32] {
+        self.0.strides.split_at(self.0.num_planes).0
+    }
+
+    /// Check if this buffer format has any vendor-specific modifiers set or is implicit/linear
+    pub fn has_modifier(&self) -> bool {
+        self.0.format.modifier != Modifier::Invalid && self.0.format.modifier != Modifier::Linear
+    }
+
+    /// Create a weak reference to this dmabuf
+    pub fn weak(&self) -> WeakDmabuf {
+        WeakDmabuf(Arc::downgrade(&self.0))
+    }
+}
+
+impl WeakDmabuf {
+    /// Try to upgrade to a strong reference of this buffer.
+    ///
+    /// Fails if no strong references exist anymore and the handle was already closed.
+    pub fn upgrade(&self) -> Option<Dmabuf> {
+        self.0.upgrade().map(Dmabuf)
+    }
+}
+
+impl Drop for DmabufInternal {
+    fn drop(&mut self) {
+        for fd in self.fds.iter() {
+            if *fd != 0 {
+                let _ = nix::unistd::close(*fd);
+            }
+        }
+    }
+}
+
+/// Buffer that can be exported as Dmabufs
+pub trait AsDmabuf {
+    /// Error type returned, if exporting fails
+    type Error;
+
+    /// Export this buffer as a new Dmabuf
+    fn export(&self) -> Result<Dmabuf, Self::Error>;
+}
+
+impl AsDmabuf for Dmabuf {
+    type Error = ();
```

Comment:
Maybe, waiting for `!` to be stabilized, we could introduce our own `Never` empty enum for use in cases where the error condition can never happen? That could simplify matches & such down the line.

Reply:
yes, probably a good idea. Do you recall other similar cases of this in smithay?

Reply:
Rust provides the error [`Infallible`](https://doc.rust-lang.org/std/convert/enum.Infallible.html) for errors that should/will never happen.

Reply:
Ah, I wasn't aware it was added in `std`, that's perfect!